config/module_input_test.go

```diff
@@ -156,3 +157,331 @@ func TestModuleInput_DecodeConfig_Error(t *testing.T) {
 		})
 	}
 }
+
+func TestModuleInputConfigs_Len(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name     string
+		i        *ModuleInputConfigs
+		expected int
+	}{
+		{
+			"nil",
+			nil,
+			0,
+		},
+		{
+			"happy_path",
+			&ModuleInputConfigs{
+				&ServicesModuleInputConfig{},
+				&ConsulKVModuleInputConfig{},
+			},
+			2,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := tc.i.Len()
+			assert.Equal(t, tc.expected, actual)
+		})
+	}
+}
+
+func TestModuleInputConfigs_Copy(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name     string
+		i        *ModuleInputConfigs
+		expected *ModuleInputConfigs
+	}{
+		{
+			"nil",
+			nil,
+			nil,
+		},
+		{
+			"empty",
+			&ModuleInputConfigs{},
+			&ModuleInputConfigs{},
+		},
+		{
+			"happy_path",
+			&ModuleInputConfigs{
+				&ServicesModuleInputConfig{
+					ServicesMonitorConfig{
+						Regexp:     String("^web.*"),
+						Datacenter: String("dc"),
+						Namespace:  String("namespace"),
+						Filter:     String("filter"),
+						CTSUserDefinedMeta: map[string]string{
+							"key": "value",
+						},
+					},
+				},
+				&ConsulKVModuleInputConfig{
+					ConsulKVMonitorConfig{
+						Path:       String("key-path"),
+						Recurse:    Bool(true),
+						Datacenter: String("dc2"),
+						Namespace:  String("ns2"),
+					},
+				},
+			},
+			&ModuleInputConfigs{
+				&ServicesModuleInputConfig{
+					ServicesMonitorConfig{
+						Regexp:     String("^web.*"),
+						Datacenter: String("dc"),
+						Namespace:  String("namespace"),
+						Filter:     String("filter"),
+						CTSUserDefinedMeta: map[string]string{
+							"key": "value",
+						},
+					},
+				},
+				&ConsulKVModuleInputConfig{
+					ConsulKVMonitorConfig{
+						Path:       String("key-path"),
+						Recurse:    Bool(true),
+						Datacenter: String("dc2"),
+						Namespace:  String("ns2"),
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := tc.i.Copy()
+			assert.Equal(t, tc.expected, actual)
+		})
+	}
+}
+
+func TestModuleInputConfigs_Merge(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name string
+		a    *ModuleInputConfigs
+		b    *ModuleInputConfigs
+		r    *ModuleInputConfigs
+	}{
+		{
+			"nil_a",
+			nil,
+			&ModuleInputConfigs{},
+			&ModuleInputConfigs{},
+		},
+		{
+			"nil_b",
+			&ModuleInputConfigs{},
+			nil,
+			&ModuleInputConfigs{},
+		},
+		{
+			"nil_both",
+			nil,
+			nil,
+			nil,
+		},
+		{
+			"empty",
+			&ModuleInputConfigs{},
+			&ModuleInputConfigs{},
+			&ModuleInputConfigs{},
+		},
+		{
+			"happy_path",
+			&ModuleInputConfigs{&ServicesConditionConfig{}},
+			&ModuleInputConfigs{&ConsulKVConditionConfig{}},
+			&ModuleInputConfigs{
+				&ServicesConditionConfig{},
+				&ConsulKVConditionConfig{},
+			},
+		},
+	}
```

Comment:
I guess the merge doesn't handle the case where two Module's of the same type are merged together...

I think this is okay though? Merge just merges the lists, it's up to the validation to catch such errors.

Reply:
The Merge() itself can handle when two of the same module_inputs are merged e.g. the below passes

```
{
	"happy_path_same_type",
	&ModuleInputConfigs{&ServicesConditionConfig{}},
	&ModuleInputConfigs{&ServicesConditionConfig{}},
	&ModuleInputConfigs{
		&ServicesConditionConfig{},
		&ServicesConditionConfig{},
	},
},
```

You're right Merge() just handles the merge, and this test case would error for Validate(). I can add this test-case

Reply:
I noticed that the test cases are "Condition" instead of "ModuleInput". Will update this as well.